src/lem/path.rs

```diff
@@ -111,77 +105,74 @@ impl LEMCTL {
         &self,
         path: &Path,
         // `map` keeps track of the updated names of variables
-        map: &mut HashMap<Var, Var>, // name -> path.name
-    ) -> Result<Self> {
+        map: &mut VarMap<Var>, // name -> path.name
+    ) -> Self {
         match self {
             LEMCTL::MatchTag(var, cases) => {
                 let mut new_cases = vec![];
                 for (tag, case) in cases {
-                    let new_case = case.deconflict(&path.push_tag(tag), &mut map.clone())?;
+                    let new_case = case.deconflict(&path.push_tag(tag), &mut map.clone());
                     new_cases.push((*tag, new_case));
                 }
-                Ok(LEMCTL::MatchTag(
-                    retrieve_one(map, var)?,
-                    IndexMap::from_iter(new_cases),
-                ))
+                LEMCTL::MatchTag(retrieve_one(map, var), IndexMap::from_iter(new_cases))
             }
             LEMCTL::MatchSymbol(var, cases, def) => {
                 let mut new_cases = vec![];
                 for (symbol, case) in cases {
-                    let new_case = case.deconflict(&path.push_symbol(symbol), &mut map.clone())?;
+                    let new_case = case.deconflict(&path.push_symbol(symbol), &mut map.clone());
                     new_cases.push((symbol.clone(), new_case));
                 }
-                Ok(LEMCTL::MatchSymbol(
-                    retrieve_one(map, var)?,
+                LEMCTL::MatchSymbol(
+                    retrieve_one(map, var),
                     IndexMap::from_iter(new_cases),
-                    Box::new(def.deconflict(&path.push_default(), &mut map.clone())?),
-                ))
+                    Box::new(def.deconflict(&path.push_default(), &mut map.clone())),
```

Comment:
Every `map.clone()` in this function is as unnecessary as it is costly, and probably erroneous to boot, because mutating the clone is probably not what you want to do.

Reply:
`map.clone()` is costly, but it is necessary (and it's what I want to do).

If `x` is defined on two alternate LEM branches, upon visiting the second we would encounter a redefinition of `x` and that would cause a panic on `VarMap::insert`. So when visiting the second branch, we must not have `x` in `map`. In other words, "deconflicting" a branch should not influence what happens on another branch.

Also, LEMs aren't too big and the number of match arms is small.

Reply:
There's a solution that can make it slightly less costly for some category of LEMs. But the algorithm will become more complex and I doubt it's worth it.

Reply:
Or you could just use some more functional/persistent environment data structure (simplest example is an alist like Lurk itself uses). You're not morally required to use a `HashMap` here.

Reply:
Let me try to summarize the issue.

When we are traversing a LEM we can i) assume it's in SSA and don't care about the consequences if it's not or ii) be more conservative and always error if we find evidence that it's not.

`VarMap` implements the second approach (hence the panics, which indicate bad plumbing). So we have that guarantee of error for `deconflict`. Now, the issue is that, while having a variable defined twice in the same path is problematic (no SSA), having variables with the same name being defined on different branches is not. So we want to prevent the breakage in these cases.

The best solution would be a data structure that use the same mechanics as the ones implemented in Lean. For example, Lean's `HashMap.insert` returns a new `HashMap` with the new pair added. This happens very efficiently internally (very few copies required).

Is there some data structure in Rust that implements this behavior? It's not the first time that I've felt the need for something like this.

Reply:
I have no comment on the implementation or licensing, but this seems to be an example of the category: https://crates.io/crates/pfds

Reply:
I think it shouldn't be too bad to be cloning these maps, since LEMs shouldn't be too big. It's hardly going to be the bottleneck. If it turns out to be a problem, we can think of other ways to deal with this.. Maybe pfds, or maybe there's a more clever way of writing this

Reply:
These are fair semantics to use. In which case, could you please add a test checking those semantics, i.e. using the same variable in two alternate branches, and checking that deconflicting it does not error / panic?

Should you want to use a semi-persistent map, https://github.com/estokes/immutable-chunkmap seems to be a good recent contender.

Reply:
There's a test case for that 👍🏼

Reply:
Oh, my bad `resolves_conflicts_of_clashing_names_in_parallel_branches`. TIL, apologies.